drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -0,0 +1,368 @@
+		break;
+	case WIFI_SECURITY_TYPE_WPA_AUTO_PERSONAL:
+		wifi_config.security = SL_WIFI_WPA2;
+		wifi_config.encryption = SL_WIFI_DEFAULT_ENCRYPTION;
+		wifi_config.credential_id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+		break;
+	/* Zephyr WiFi shell doesn't specify how to pass credential for these
+	 * key managements.
+	 */
+	case WIFI_SECURITY_TYPE_WEP: /* SL_WIFI_WEP/SL_WIFI_WEP_ENCRYPTION */
+	case WIFI_SECURITY_TYPE_EAP: /* SL_WIFI_WPA2_ENTERPRISE/<various> */
+	case WIFI_SECURITY_TYPE_WAPI:
+	default:
+		return -ENOTSUP;
+	}
+
+	if (params->band != WIFI_FREQ_BAND_UNKNOWN && params->band != WIFI_FREQ_BAND_2_4_GHZ) {
+		return -ENOTSUP;
+	}
+
+	if (params->psk_length) {
+		sl_net_set_credential(SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID, SL_NET_WIFI_PSK,
+				      params->psk, params->psk_length);
+	}
+
+	if (params->sae_password_length) {
+		sl_net_set_credential(SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID, SL_NET_WIFI_PSK,
+				      params->sae_password, params->sae_password_length);
+	}
+
+	if (params->channel != WIFI_CHANNEL_ANY) {
+		wifi_config.channel.channel = params->channel;
+	}
+
+	wifi_config.ssid.length = params->ssid_length,
+	memcpy(wifi_config.ssid.value, params->ssid, params->ssid_length);
+
+	ret = sl_wifi_connect(SL_WIFI_CLIENT_INTERFACE, &wifi_config, 0);
+	if (ret != SL_STATUS_IN_PROGRESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int siwx91x_disconnect(const struct device *dev)
+{
+	struct siwx91x_dev *sidev = dev->data;
+	int ret;
+
+	ret = sl_wifi_disconnect(SL_WIFI_CLIENT_INTERFACE);
+	if (ret) {
+		return -EIO;
+	}
+	if (IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_NATIVE)) {
+		net_eth_carrier_off(sidev->iface);
+	}
+	sidev->state = WIFI_STATE_INACTIVE;
+	return 0;
+}
+
+static void siwx91x_report_scan_res(struct siwx91x_dev *sidev, sl_wifi_scan_result_t *result,
+				    int item)
+{
+	static const struct {
+		int sl_val;
+		int z_val;
+	} security_convert[] = {
+		{ SL_WIFI_OPEN,            WIFI_SECURITY_TYPE_NONE    },
+		{ SL_WIFI_WEP,             WIFI_SECURITY_TYPE_WEP     },
+		{ SL_WIFI_WPA,             WIFI_SECURITY_TYPE_WPA_PSK },
+		{ SL_WIFI_WPA2,            WIFI_SECURITY_TYPE_PSK     },
+		{ SL_WIFI_WPA3,            WIFI_SECURITY_TYPE_SAE     },
+		{ SL_WIFI_WPA3_TRANSITION, WIFI_SECURITY_TYPE_SAE     },
+		{ SL_WIFI_WPA_ENTERPRISE,  WIFI_SECURITY_TYPE_EAP     },
+		{ SL_WIFI_WPA2_ENTERPRISE, WIFI_SECURITY_TYPE_EAP     },
+	};
+	struct wifi_scan_result tmp = {
+		.channel = result->scan_info[item].rf_channel,
+		.rssi = result->scan_info[item].rssi_val,
+		.ssid_length = strlen(result->scan_info[item].ssid),
+		.mac_length = sizeof(result->scan_info[item].bssid),
+		.security = WIFI_SECURITY_TYPE_UNKNOWN,
+		.mfp = WIFI_MFP_UNKNOWN,
+		/* FIXME: fill .mfp, .band and .channel */
+	};
+	int i;
+
+	memcpy(tmp.ssid, result->scan_info[item].ssid, tmp.ssid_length);
+	memcpy(tmp.mac, result->scan_info[item].bssid, tmp.mac_length);
+	for (i = 0; i < ARRAY_SIZE(security_convert); i++) {
```

Comment:
This certainly works but there is also `ARRAY_FOR_EACH(security_convert, i)` macro could be used here.

Reply:
Also fixed the occurrences in siwx91x_wifi_socket.c.